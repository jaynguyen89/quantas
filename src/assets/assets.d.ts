declare module '*.png' {
    const path: string;
    export default path;
}

declare module 'blob-polyfill';
