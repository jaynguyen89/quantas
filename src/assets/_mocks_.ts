import { IApiHotel, IHotel } from 'src/commons/interfaces';
import { CancellationType, ImageType, PromotionType, RatingType } from 'src/commons/enums';

export const mockedApiHotelData: {
    results: Array<IApiHotel>,
} = {
    results: [
        {
            id: '123',
            property: {
                propertyId: '111',
                title: 'First Hotel',
                address: ['1 What Street', 'Where 3333'],
                previewImage: {
                    url: 'https://photo1.png',
                    caption: 'Photo1',
                    imageType: 'PRIMARY',
                },
                rating: {
                    ratingValue: 3,
                    ratingType: 'self',
                },
            },
            offer: {
                promotion: {
                    title: 'Promo1',
                    type: 'MEMBER',
                },
                name: 'Promo1',
                displayPrice: {
                    amount: 111,
                    currency: 'AU',
                },
                savings: {
                    amount: 11,
                    currency: 'AU',
                },
                cancellationOption: {
                    cancellationType: 'NOT_REFUNDABLE',
                },
            },
        },
        {
            id: '456',
            property: {
                propertyId: '222',
                title: 'Second Hotel',
                address: ['2 Which Way', 'Anywhere 2222'],
                previewImage: {
                    url: 'https://photo2.png',
                    caption: 'Photo2',
                    imageType: 'PRIMARY',
                },
                rating: {
                    ratingValue: 4,
                    ratingType: 'star',
                },
            },
            offer: {
                promotion: {
                    title: 'Promo2',
                    type: 'MEMBER',
                },
                name: 'Promo2',
                displayPrice: {
                    amount: 222,
                    currency: 'US',
                },
                savings: null,
                cancellationOption: {
                    cancellationType: 'FREE_CANCELLATION',
                },
            },
        },
        {
            id: '789',
            property: {
                propertyId: '333',
                title: 'Third Hotel',
                address: ['3 Where Road', 'Somewhere 1111', 'Australia'],
                previewImage: {
                    url: 'https://photo3.png',
                    caption: 'Photo3',
                    imageType: 'PRIMARY',
                },
                rating: {
                    ratingValue: 5,
                    ratingType: 'self',
                },
            },
            offer: {
                promotion: {
                    title: 'Promo3',
                    type: 'MEMBER',
                },
                name: 'Promo3',
                displayPrice: {
                    amount: 333,
                    currency: 'VND',
                },
                savings: {
                    amount: 33,
                    currency: 'VND',
                },
                cancellationOption: {
                    cancellationType: 'FREE_CANCELLATION',
                },
            },
        },
    ],
};

export const mockedHotelData: Array<IHotel> = [
    {
        id: '123',
        name: 'First Hotel',
        address: '1 What Street, Where 3333',
        image: {
            url: 'https://photo1.png',
            caption: 'Photo1',
            type: ImageType.PRIMARY,
        },
        rating: {
            value: 3,
            type: RatingType.SELF,
        },
        pricing: {
            name: 'Promo1',
            price: {
                value: 111,
                currency: 'AU',
            },
            promotion: {
                title: 'Promo1',
                type: PromotionType.MEMBER,
            },
            savings: {
                value: 11,
                currency: 'AU',
            },
            cancellationType: CancellationType.FREE_CANCELLATION,
        },
    },
    {
        id: '456',
        name: 'Second Hotel',
        address: '2 Which Way, Where 2222',
        image: {
            url: 'https://photo2.png',
            caption: 'Photo2',
            type: ImageType.PRIMARY,
        },
        rating: {
            value: 4,
            type: RatingType.STAR,
        },
        pricing: {
            name: 'Promo2',
            price: {
                value: 222,
                currency: 'US',
            },
            promotion: {
                title: 'Promo2',
                type: PromotionType.MEMBER,
            },
            cancellationType: CancellationType.FREE_CANCELLATION,
        },
    },
    {
        id: '789',
        name: 'Third Hotel',
        address: '3 Where Road, Somewhere 1111, Australia',
        image: {
            url: 'https://photo3.png',
            caption: 'Photo3',
            type: ImageType.PRIMARY,
        },
        rating: {
            value: 5,
            type: RatingType.SELF,
        },
        pricing: {
            name: 'Promo3',
            price: {
                value: 333,
                currency: 'US',
            },
            promotion: {
                title: 'Promo3',
                type: PromotionType.MEMBER,
            },
            savings: {
                value: 33,
                currency: 'US',
            },
            cancellationType: CancellationType.FREE_CANCELLATION,
        },
    },
];