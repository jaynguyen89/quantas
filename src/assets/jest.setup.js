import { TextEncoder, TextDecoder, ReadableStream } from 'node:util';

global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;
global.ReadableStream = ReadableStream;