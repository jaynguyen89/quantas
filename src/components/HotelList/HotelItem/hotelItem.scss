@use 'src/commons/variables' as vars;

.hotel-item {
    width: 100%;
    display: flex;
    flex-direction: row;
    margin-bottom: vars.$xxtiny;

    .hotel-img {
        position: relative;

        img {
            width: 150px;
            height: 125px;
        }

        p {
            font-size: vars.$xh10;
            font-weight: vars.$thick;
            color: vars.$red;
            background-color: vars.$white;
            position: absolute;
            top: vars.$sh10;
            padding: vars.$micro;
            border: vars.$xmicro solid vars.$lightGrey;
        }
    }

    .hotel-info {
        width: 100%;
        margin-left: vars.$h8;
        padding: vars.$micro 0;
        font-weight: vars.$thick;
        font-size: vars.$xh10;
        position: relative;
        border-top: vars.$xmicro solid vars.$lightGrey;
        display: flex;
        flex-direction: row;
        justify-content: space-between;

        .general-info {
            h6 {
                font-size: vars.$mh9;
                font-weight: vars.$regular;
                margin: 0;
            }

            p {
                font-weight: vars.$light;
                color: vars.$grey;
                margin: 0 0 vars.$xxtiny;
            }

            a {
                color: vars.$red;
                font-weight: vars.$light;
            }

            span:not(h6 span) {
                font-weight: vars.$light;
                color: vars.$green;
                position: absolute;
                bottom: 0;
            }
        }

        .price-info {
            align-content: center;
            text-align: right;
            font-size: vars.$sh9;
            font-weight: vars.$light;

            .pricing {
                width: vars.$xh1;

                p {
                    font-size: vars.$xxtiny;
                    margin: 0 0 vars.$sh10;
                }

                h5 {
                    font-weight: vars.$light;
                    font-size: vars.$h8;
                    margin: 0;

                    span {
                        font-size: vars.$h9;
                        vertical-align: top;
                    }
                }

                b {
                    margin: 0;
                    color: vars.$red;
                    font-weight: vars.$light;

                    span {
                        vertical-align: top;
                    }
                }
            }
        }
    }
}