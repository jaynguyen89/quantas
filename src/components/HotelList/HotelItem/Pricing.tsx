import React from 'react';
import 'src/components/HotelList/HotelItem/hotelItem.scss';
import { IOffer } from 'src/commons/interfaces';

type TPricingProps = {
    pricing: IOffer,
};

const Pricing = ({
    pricing,
}: TPricingProps) => {
    return (
        <div className='pricing'>
            <p>1 night total ({pricing.price.currency})</p>
            <h5><span>$</span>{pricing.price.value}</h5>

            {pricing.savings && (
                <b>Save ${pricing.savings.value}<span>~</span></b>
            )}
        </div>
    );
};

export default Pricing;