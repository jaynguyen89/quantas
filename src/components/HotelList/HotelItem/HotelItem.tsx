import React from 'react';
import 'src/components/HotelList/HotelItem/hotelItem.scss';
import Rating from 'src/components/HotelList/Rating/Rating';
import Pricing from 'src/components/HotelList/HotelItem/Pricing';
import { IHotel } from 'src/commons/interfaces';
import { CancellationType } from 'src/commons/enums';

type THotelItemProps = {
    hotel: IHotel,
};

const HotelItem = ({
    hotel,
}: THotelItemProps) => {
    const cancellationText =
        hotel.pricing.cancellationType === CancellationType.FREE_CANCELLATION
            ? 'Free cancellation'
            : null;

    return (
        <div className='hotel-item'>
            <div className='hotel-img'>
                <img src={hotel.image.url} alt={hotel.image.caption} />
                <p>{hotel.pricing.promotion.title}</p>
            </div>

            <div className='hotel-info'>
                <div className='general-info'>
                    <h6>
                        <span id='hotel-name'>{hotel.name}</span>
                        <Rating {...hotel.rating} />
                    </h6>
                    <p>{hotel.address}</p>
                    <a href='/'>{hotel.pricing.name}</a>
                    <br />
                    {cancellationText && (
                        <span>{cancellationText}</span>
                    )}
                </div>

                <div className='price-info'>
                    <Pricing pricing={hotel.pricing} />
                </div>
            </div>
        </div>
    );
};

export default HotelItem;