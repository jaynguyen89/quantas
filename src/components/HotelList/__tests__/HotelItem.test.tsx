import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import toJson from 'enzyme-to-json';
import { render } from '@testing-library/react';
import HotelItem from 'src/components/HotelList/HotelItem/HotelItem';
import { mockedHotelData } from 'src/assets/_mocks_';
import { getBy } from 'src/utilities/jest.utils';

configure({ adapter: new Adapter() });

describe('HotelItem.tsx', () => {
    it('renders correctly', () => {
        const wrapper = shallow(<HotelItem hotel={mockedHotelData[0]} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('contains the necessary elements', () => {
        const data = mockedHotelData[0];
        const wrapper = render(<HotelItem hotel={data} />);

        const content = wrapper.container.getElementsByClassName('hotel-item');
        expect(content.length).toEqual(1);
        expect(content[0]).toBeDefined();

        const imgContainer = content[0].getElementsByClassName('hotel-img');
        expect(imgContainer.length).toEqual(1);
        expect(imgContainer[0]).toBeDefined();

        const imgEl = imgContainer[0].getElementsByTagName('img');
        expect(imgEl.length).toEqual(1);
        expect(imgEl[0]).toBeDefined();
        expect(imgEl[0].src).toContain(data.image.url);

        const promoEl = imgContainer[0].getElementsByTagName('p');
        expect(promoEl.length).toEqual(1);
        expect(promoEl[0]).toBeDefined();
        expect(promoEl[0].innerHTML).toEqual(data.pricing.promotion.title);

        const infoContainer = content[0].getElementsByClassName('hotel-info');
        expect(infoContainer.length).toEqual(1);
        expect(infoContainer[0]).toBeDefined();

        const generalInfoContainer = infoContainer[0].getElementsByClassName('general-info');
        expect(generalInfoContainer.length).toEqual(1);
        expect(generalInfoContainer[0]).toBeDefined();

        const h6El = generalInfoContainer[0].getElementsByTagName('h6');
        expect(h6El.length).toEqual(1);
        expect(h6El[0]).toBeDefined();

        const h6Title = getBy(wrapper.container, 'id', 'hotel-name');
        expect(h6Title).toBeDefined();
        expect(h6Title!.innerHTML).toEqual(data.name);

        const ratingContainer = generalInfoContainer[0].getElementsByClassName('rating');
        expect(ratingContainer.length).toEqual(1);
        expect(ratingContainer[0]).toBeDefined();

        generalInfoContainer[0]!.removeChild(h6El[0]);
        const addressEl = generalInfoContainer[0].getElementsByTagName('p');
        expect(addressEl.length).toEqual(1);
        expect(addressEl[0]).toBeDefined();
        expect(addressEl[0]!.innerHTML).toEqual(data.address);

        const linkEl = generalInfoContainer[0].getElementsByTagName('a');
        expect(linkEl.length).toEqual(1);
        expect(linkEl[0]).toBeDefined();
        expect(linkEl[0].innerHTML).toEqual(data.pricing.name);

        const cancellationEl = generalInfoContainer[0].getElementsByTagName('span');
        expect(cancellationEl.length).toEqual(1);
        expect(cancellationEl[0]).toBeDefined();
        expect(cancellationEl[0].innerHTML).toEqual('Free cancellation');

        const pricingInfoContainer = infoContainer[0].getElementsByClassName('price-info');
        expect(pricingInfoContainer.length).toEqual(1);
        expect(pricingInfoContainer[0]).toBeDefined();

        const pricingContainer = pricingInfoContainer[0].getElementsByClassName('pricing');
        expect(pricingContainer.length).toEqual(1);
        expect(pricingContainer[0]).toBeDefined();
    });
});