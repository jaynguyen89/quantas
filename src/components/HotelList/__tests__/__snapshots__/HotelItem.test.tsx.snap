// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelItem.tsx renders correctly 1`] = `
<div
  className="hotel-item"
>
  <div
    className="hotel-img"
  >
    <img
      alt="Photo1"
      src="https://photo1.png"
    />
    <p>
      Promo1
    </p>
  </div>
  <div
    className="hotel-info"
  >
    <div
      className="general-info"
    >
      <h6>
        <span
          id="hotel-name"
        >
          First Hotel
        </span>
        <Rating
          type="SELF"
          value={3}
        />
      </h6>
      <p>
        1 What Street, Where 3333
      </p>
      <a
        href="/"
      >
        Promo1
      </a>
      <br />
      <span>
        Free cancellation
      </span>
    </div>
    <div
      className="price-info"
    >
      <Pricing
        pricing={
          Object {
            "cancellationType": "FREE_CANCELLATION",
            "name": "Promo1",
            "price": Object {
              "currency": "AU",
              "value": 111,
            },
            "promotion": Object {
              "title": "Promo1",
              "type": "MEMBER",
            },
            "savings": Object {
              "currency": "AU",
              "value": 11,
            },
          }
        }
      />
    </div>
  </div>
</div>
`;
